src/main/kotlin/io/github/mojira/arisa/ModuleExecutor.kt

```diff
@@ -0,0 +1,333 @@
+    private val revokeConfirmationModule: RevokeConfirmationModule
+
+    private val ticketCache = mutableListOf<String>()
+    private val ticketTimer = Timer("RemoveCachedTicket", true)
+
+    private fun executeModule(
+        moduleConfig: Arisa.Modules.ModuleConfigSpec,
+        queryCache: MutableMap<String, List<Issue>>,
+        processedIssues: MutableMap<String, Boolean>,
+        executeModule: (Issue) -> Pair<String, Either<ModuleError, ModuleResponse>>
+    ) {
+        val projects = config[Arisa.Issues.projects]
+            .filter { config[moduleConfig.whitelist] == null || config[moduleConfig.whitelist]!!.contains(it) }
+            .joinToString(",")
+        val resolutions = config[moduleConfig.resolutions].joinToString(",") { "\"$it\"" }
+        val cachedTickets = if (ticketCache.isEmpty()) "" else "AND key not in (${ticketCache.joinToString(",")})"
+        val combinedJql = "project in ($projects) $cachedTickets AND resolution in ($resolutions) AND (${config[moduleConfig.jql]})"
+
+        val issues = queryCache[combinedJql] ?: jiraClient
+            .searchIssues(combinedJql)
+            .issues
+            .map { jiraClient.getIssue(it.key, "*all", "changelog") } // Get issues again to retrieve all fields
+            .filter { issue ->
+                // Ignore issues where last action was a resolve
+                val latestChange = issue.changeLog.entries.lastOrNull()
+
+                latestChange == null || // There is actually a entry
+                    !latestChange.items.any { it.field == "resolution" } || // It was a transition
+                    latestChange.author.name == config[Arisa.Credentials.username] || // The transition was not done by the bot
+                    (issue.comments.isNotEmpty() && issue.comments.last().updatedDate > latestChange.created) // And there is no comment posted after that
+            }
+        queryCache[combinedJql] = issues
+
+        issues
+            .map { it.key to executeModule(it) }
+            .forEach { (issue, response) ->
+                response.second.fold({
+                    processedIssues.putIfAbsent(issue, false)
+                    when (it) {
+                        is OperationNotNeededModuleResponse -> log.info("[RESPONSE] [$issue] [${response.first}] Operation not needed")
+                        is FailedModuleResponse -> for (exception in it.exceptions) {
+                            log.error("[RESPONSE] [$issue] [${response.first}] Failed", exception)
+                        }
+                    }
+                }, {
+                    processedIssues[issue] = true
+                    log.info("[RESPONSE] [$issue] [${response.first}] Successful")
+                })
+            }
+    }
+
+    fun execute() {
+        // Cache issues returned by a query to avoid searching the same query for different modules
+        val queryCache = mutableMapOf<String, List<Issue>>()
```

Comment:
[minor] I would open a issue to move all the cache stuff to its own class just to clean the executors. Will do in my own branch after this is merged :D

Reply:
#77